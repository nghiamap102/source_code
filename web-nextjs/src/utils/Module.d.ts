import { IFile } from './modules/File.d'
import { ICategoryItem } from './modules/Category.d'
import { IPagination, IPaginations, IPaginationOption, IPagitaionAction } from './modules/Pagination.d'
import { ICarouselList, IContentType, ICarouselItem, IImageProps } from './modules/Image.d'
import { IFilterBy, IFilterData, IFilterItem, IItemCheckbox, ISearch, ISelectedTag, IArticles } from './modules/Filter'

export {
  ICarouselList,
  IContentType,
  ICarouselItem,
  IImageProps,
  IFile,
  ICategoryItem,
  IPagination,
  IPaginations,
  IPaginationOption,
  IPagitaionAction,
  IFilterBy,
  IFilterData,
  IFilterItem,
  IItemCheckbox,
  ISearch,
  ISelectedTag,
  IArticles
}
