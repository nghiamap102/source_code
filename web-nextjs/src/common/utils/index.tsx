import _ from 'lodash'

export const formatMoney = (value: any) => `${value}`.toString().replace(/\B(?=(\d{3})+(?!\d))/g, '.') + 'Ä‘'

export const formatTime = (time: any) => {
  return new Date(time).toLocaleTimeString('vi', {
    // day: "2-digit",
    hour: '2-digit',
    minute: '2-digit'
  })
}

export const formatDate = (date: any) => {
  return date.replace(/(\d{4})-(\d\d)-(\d\d)/, '$3-$2-$1')
}

export const isEmail = (emailAddress: string) => {
  return emailAddress.match(/^\w+([\.-]?\w+)*@\w+([\.-]?\w+)*(\.\w{2,3})+$/)
}

export const formatQuery = (queryString: string) => {
  return JSON.parse('{"' + decodeURI(queryString).replace(/"/g, '\\"').replace(/&/g, '","').replace(/=/g, '":"') + '"}')
}

export const doubleNumber = (num: number): number => {
  return num * 2
}

export const getLineClamp = (countLine: number) => {
  return {
    display: '-webkit-box',
    '-webkit-line-clamp': `${countLine}`,
    '-webkit-box-orient': 'vertical',
    overflow: 'hidden',
    'text-overflow': 'ellipsis',
    width: '100%',
    'max-width': '100%',
    'word-break': 'break-word'
  }
}

export const funcWithDebounce = _.debounce((func) => func(), 500)
