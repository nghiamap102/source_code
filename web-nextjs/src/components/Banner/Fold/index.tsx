import { Image } from '@chakra-ui/react'
import { StyleProps } from '@chakra-ui/system'

import React, { FC } from 'react'

interface Props {
  customStyles?: StyleProps
}

const Fold: FC<Props> = ({ customStyles }) => {
  return (
    <Image
      src='/images/fold-banner.png'
      bg='main.300'
      w='100%'
      h={{ base: '13.03px', md: '50px' }}
      mb={{ base: '-13.03px', md: '-50px' }}
      {...customStyles}
      alt='fold-banner'
    />
  )
}

export default Fold
