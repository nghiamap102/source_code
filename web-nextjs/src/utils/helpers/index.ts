import { FormikValues } from 'formik'
import { IntlMessageFormat } from 'intl-messageformat'
import get from 'lodash/get'
import has from 'lodash/has'
import keys from 'lodash/keys'
import { ICarouselItem, ICarouselList } from '../Module.d'

const isNonEmptyArray = (items: Array<any>) => Array.isArray(items) && items.length > 0

const stringToHTML = (str: string) => {
  const e = document.createElement('div')
  e.innerHTML = str
  return e.childNodes.length === 0 ? '' : e.childNodes[0].nodeValue
}

const onErrorImage = (w: number = 150, h?: number): string => {
  const resource = 'https://via.placeholder.com'
  const imgSize = h ? `${w}x${h}` : w

  return `${resource}/${imgSize}`
}

const getBanner = (data: ICarouselItem) => {
  const desktop = data.field_ci_image_desktop
  const mobile = data.field_ci_image_mobile

  return {
    desktop: {
      url: desktop?.url,
      absolute_url: desktop?.absolute_url
    },
    mobile: {
      url: mobile?.url,
      absolute_url: mobile?.absolute_url
    }
  }
}

const mappingDataBanner = (data: ICarouselList) => {
  const carouselItem = get(data, 'field_carousel_items', [])
  if (!isNonEmptyArray(carouselItem)) return []

  return carouselItem.map((data: ICarouselItem) => {
    const images = getBanner(data)
    const { title, uid, created } = data
    return {
      uid,
      title,
      created,
      images: {
        desktop: images?.desktop.absolute_url || images?.desktop.url,
        mobile: images?.mobile.absolute_url || images?.mobile.url
      },
      fallback: {
        desktop: onErrorImage(1440, 900),
        mobile: onErrorImage(300, 400)
      }
    }
  })
}

export type validationFieldProps = {
  validator: (...params: any) => boolean
  code: string
  codeOptions?: {
    [key: string]: any
  }
}

export interface IValidations {
  [key: string]: validationFieldProps[]
}

const checkValueError = (validations: IValidations) => (values: FormikValues, props?: any) => {
  const error: { [key: string]: any } = {}
  let checkValidate = false
  keys(validations).forEach((path: string) => {
    const pathValue = get(values, path)
    const isExistingKey = has(values, path)
    if (!isExistingKey) {
      // tslint:disable-next-line:no-console
      console && console.error(`The field ${path} does not existing on the form`)
    }
    for (let i = 0; i < validations[path].length; i += 1) {
      const pathItem = validations[path][i] ?? {}
      checkValidate = pathItem.validator(pathValue, values, props)
      if (!checkValidate) {
        const code = pathItem.code
        const codeOptions = pathItem.codeOptions ?? {}
        const codeOptionLength = Object.entries(codeOptions)?.length
        const newMessageFormat =
          codeOptionLength > 0
            ? new IntlMessageFormat(code, 'en').format({ ...codeOptions })
            : new IntlMessageFormat(code, 'en').format()
        error[path] = newMessageFormat
        return error
      }
    }
  })

  return error
}

const paginationWithDots = (c: any, m: any) => {
  const current = c
  const last = m
  const delta = 2
  const left = current - delta
  const right = current + delta + 1
  const range = []
  const rangeWithDots = []
  let l

  // eslint-disable-next-line no-plusplus
  for (let i = 0; i <= last; i++) {
    if (i === 0 || i === last || (i >= left && i < right)) {
      range.push(i)
    }
  }

  for (const i of range) {
    if (l) {
      if (i - l === 2) {
        rangeWithDots.push(l + 1)
      } else if (i - l !== 1) {
        rangeWithDots.push('...')
      }
    }
    rangeWithDots.push(i)
    l = i
  }

  return rangeWithDots
}

const omit = (obj: any, keyss: any) => Object.fromEntries(Object.entries(obj).filter(([k]) => !keyss.includes(k)))

export {
  isNonEmptyArray,
  stringToHTML,
  getBanner,
  checkValueError,
  onErrorImage,
  paginationWithDots,
  mappingDataBanner,
  omit
}
