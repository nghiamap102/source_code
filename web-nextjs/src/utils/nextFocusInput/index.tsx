interface interFaceNextFocusInput {
  maxLength: number
  value: any
  name: string
}

const nextFocusInput = (e: interFaceNextFocusInput) => {
  const { maxLength, value, name } = e
  const [fieldName, fieldIndex] = name.split('_')
  if (maxLength === value.length) {
    const nextSibling = document.querySelector(`input[name=${fieldName}_${+fieldIndex + 1}]`) as HTMLElement
    if (nextSibling !== null) {
      nextSibling.focus()
    }
  }
}

export default nextFocusInput
