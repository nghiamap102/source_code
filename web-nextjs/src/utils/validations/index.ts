import isEmpty from 'lodash/isEmpty'

const DEFAULT_MAX_NUMBER_INPUT = 99999999

export const validateRequired = (value: string | number) => {
  let error = true
  if (isEmpty(value) && isEmpty(value && value.toString())) {
    error = false
  }
  return error
}

export const validateRequiredCheckbox = (value: string | boolean) => {
  if (!value) return false
  return true
}

const emailFormat =
  /^(([^<>()[\]\\.,;:\s@"]+(\.[^<>()[\]\\.,;:\s@"]+)*)|(".+"))@((\[[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}\])|(([a-zA-Z\-0-9]+\.)+[a-zA-Z]{2,}))$/
const format = /[!#$%^&*()+\\=\\[\]{};':"\\|,<>\\/?]+/
const formatSymbols = /[`!@#$%^&*()_+\-=\\[\]{};':"\\|,.<>\\/?~]/
// eslint-disable-next-line
const notUnicodeFormat = /[^\u0000-\u00ff]/
const onlyContainCharacterRegex = /^[a-zA-Z@$!%*#?&\040]+$/
const containAtLeastNumberRegex = /(?=.*\d)/

const passwordFormat = /^(?=.*[a-z])(?=.*[A-Z])(?=.*\d)[a-zA-Z\d]{6,}$/

const passwordSpecialFormat = /^(?=.*[a-z])(?=.*[A-Z])(?=.*\d)(?=.*[@$!%*?&])[A-Za-z\d@$!%*?~#&]{6,}$/

export const validFileExtensions = ['.jpg', '.jpeg', '.gif', '.png']

export const validateEmail = (email: string) => {
  const lowerCase = String(email).toLowerCase()
  if (!emailFormat.test(lowerCase) || notUnicodeFormat.test(lowerCase)) {
    return false
  }
  return true
}

export const validatePassword = (password: string) => {
  if (!passwordFormat.test(password) && !passwordSpecialFormat.test(password)) return false
  return true
}

export const validateNotSymbols = (text: string) => {
  const lowerCase = String(text).toLowerCase()
  if (!formatSymbols.test(lowerCase) || notUnicodeFormat.test(lowerCase)) {
    return true
  }
  return false
}

export const validateMobile = (mobile: string) => {
  const lowerCase = String(mobile).toLowerCase()
  if (isNaN(Number(mobile))) {
    if (!Number(lowerCase.substring(2)) || lowerCase[1] === '+') {
      return false
    }
  }
  return true
}

// export const validateCardNumber = (cardNumber: string) => {
//   const validationCardTypes = JSON.parse(localStorage.getItem('cardValidation'))
//   const cardNumberFormat = cardNumber.replaceAll(' ', '')
//   let flag = false
//   if (validationCardTypes.length > 0) {
//     validationCardTypes.forEach(item => {
//       const regex = new RegExp(item.regex)
//       if (cardNumberFormat.match(regex)) {
//         flag = true
//       }
//     })
//   }
//   if (flag) return true
//   return false
// }

export const validateMonth = (month: string) => {
  if (Number(month) < 1 || Number(month) > 12) {
    return false
  }
  const getYear: any = document.querySelector('[data-encrypt="year"]')
  const year: string = getYear?.value || ''
  if (!year) return true
  const currentMonth = new Date().getMonth() + 1
  const currentYear = new Date().getFullYear()

  if (Number(year) < currentYear || (Number(year) === currentYear && Number(month) < currentMonth)) return false
  return true
}

export const validateYear = (year: string) => {
  const currentMonth = new Date().getMonth() + 1
  const currentYear = new Date().getFullYear()
  const getMonth: any = document.querySelector('[data-encrypt="month"]')
  const month = getMonth || ''
  if (Number(year) < currentYear || Number(year) > currentYear + 10) {
    return false
  }
  if (!month) return true
  if (Number(year) < currentYear || (Number(year) === currentYear && Number(month) < currentMonth)) return false
  return true
}

export const validateEmailOrPhone = (data: string) => {
  const lowerCase: string = String(data).toLowerCase()
  if (isNaN(+lowerCase)) {
    if (format.test(lowerCase) || !emailFormat.test(lowerCase) || notUnicodeFormat.test(lowerCase)) {
      return false
    }
  }
  if (lowerCase.length < 8) {
    return false
  }
  if (validateMobile(lowerCase) && lowerCase.length > 15) {
    return false
  }
  return true
}

export const validateRetype = (values: { confirmPassword: string; password: string }) => {
  let error = true
  if (values.confirmPassword && values.confirmPassword !== values.password) {
    error = false
  }
  return error
}

export const validateMinLength = (minLength: number) => (value: string) => String(value).length >= minLength

export const validateMaxLength = (maxLength: number) => (value: string) => String(value).length <= maxLength

export const validateContainLeastCharacter = (value: string) => {
  const character = value.replace(/[0-9]/g, '')
  return onlyContainCharacterRegex.test(character)
}

export const validateContainLeastNumber = (value: string) => containAtLeastNumberRegex.test(value)

export const validateNumberLarger0 = (value: string) => {
  if (value) {
    const regexp = /^\d+$/
    let valid = true
    if (!regexp.test(value)) {
      valid = false
    }
    if (Number(value) <= 0) {
      valid = false
    }
    return valid
  }

  return true
}

export const validateNumberLargerMaxNumber = (value: string) => {
  if (value) {
    const regexp = /^\d+$/
    let valid = true
    if (!regexp.test(value)) {
      valid = false
    }
    if (Number(value) > DEFAULT_MAX_NUMBER_INPUT) {
      valid = false
    }
    return valid
  }

  return true
}
export const validateContainAtLeastOneCapitalCharacter = (value: string) => {
  if (value) {
    const regex = /(?=.*[A-Z])/
    return regex.test(value)
  }
  return true
}

export const validateURLWebsite = (value: string) => {
  // eslint-disable-next-line
  const regexp = /^(http|https|ftp):\/\/[a-z0-9]+([\\-\\.]{1}[a-z0-9]+)*\.[a-z]{2,5}(:[0-9]{1,5})?(\/.*)?$/i
  if (!value) return false
  if (regexp.test(value)) {
    return true
  }
  return false
}

export const validStringWithoutBlank = (value: string) => {
  if (value.indexOf(' ') > 1) {
    return false
  }
  return true
}

export const validStringWithoutUppercase = (value: string) => {
  const regexp = /[A-Z]+/g
  if (!value) return false
  if (regexp.test(value)) {
    return false
  }
  return true
}

export const validateKeyProject = (value: string) => {
  const regexp = /^[A-Z][A-Z0-9]+/g
  if (!value) return false
  if (regexp.test(value)) {
    return true
  }
  return false
}
export const validateImageType = (fileName: string) => {
  const regexp = /\.(gif|jpe?g|tiff?|png|webp|bmp)$/i
  if (regexp.test(fileName)) {
    return true
  }
  return false
}

export const validateVideoType = (fileName: string) => {
  const regexp = /\.(m4v|avi|mpg|mp4)$/i
  if (regexp.test(fileName)) {
    return true
  }
  return false
}

export const validStringWithoutSpecialCharacter = (value: string) => {
  const regexp = /`|~|!|@|#|\$|%|\^|&|\*|\(|\)|\+|=|\[|\{|\]|\}|\||\\|'|<|,|\.|>|\?|\/|"|;|:/g
  if (!value) return false
  if (regexp.test(value)) {
    return false
  }
  return true
}
