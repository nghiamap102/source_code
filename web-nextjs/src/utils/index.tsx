import { isNonEmptyArray, onErrorImage, getBanner, paginationWithDots, mappingDataBanner } from './helpers'
import nextFocusInput from './nextFocusInput'
import {
  validateContainLeastCharacter,
  validateContainLeastNumber,
  validateEmail,
  validateMaxLength,
  validateMinLength,
  validateMonth,
  validatePassword,
  validateRetype,
  validateYear
} from './validations'

export {
  nextFocusInput,
  validateEmail,
  validateContainLeastCharacter,
  validateMaxLength,
  validateMinLength,
  validateMonth,
  validateYear,
  validatePassword,
  validateRetype,
  validateContainLeastNumber,
  isNonEmptyArray,
  onErrorImage,
  getBanner,
  paginationWithDots,
  mappingDataBanner
}
