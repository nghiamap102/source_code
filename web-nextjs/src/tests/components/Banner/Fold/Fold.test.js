/* eslint-disable no-undef */
import Fold from 'src/components/Banner/Fold'
import { testSnapshots } from 'src/tests/utils/getSnapshot'

describe('<Banner />', () => {
  testSnapshots(Fold, [
    {
      props: {},
      description: 'render Banner Fold default'
    }
  ])
})
