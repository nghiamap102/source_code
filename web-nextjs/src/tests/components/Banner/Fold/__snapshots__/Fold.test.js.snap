// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Banner /> snapshots render Banner Fold default 1`] = `
<div>
  <img
    alt="fold-banner"
    class="chakra-image css-xz30w5"
    src="/images/fold-banner.png"
  />
</div>
`;
